.github/workflows/publish.yml

```diff
@@ -11,34 +11,73 @@ on:
     paths:
       - "src/**"
       - ".github/workflows/publish.yml"
+  repository_dispatch:
+    types: [create-release]
+    client_payload:
+      release_tag:
+        required: true
 
 jobs:
+  create_release:
+    if: ${{ github.event_name == 'repository_dispatch' }}
+    runs-on: ubuntu-latest
+    steps:
+      - name: checkout
+        uses: Brightspace/third-party-actions@actions/checkout
+      - name: Create release
+        env:
+          RELEASE_TAG: ${{ github.event.client_payload.release_tag }}
+        run: |
+          gh release create "$RELEASE_TAG" --draft `
+            --title "Release $RELEASE_TAG" `
+            --notes "Release notes for version $RELEASE_TAG"
+
   build:
     strategy:
       matrix:
         include:
           - machine: windows-latest
             platform: win
+            file_name: D2L.Bmx.exe
           - machine: macos-latest
             platform: osx
+            file_name: D2L.Bmx
     runs-on: ${{ matrix.machine }}
     timeout-minutes: 10
+    env:
+      PLATFORM: ${{ matrix.platform }}
     steps:
       - name: checkout
         uses: Brightspace/third-party-actions@actions/checkout
       - name: publish
         shell: pwsh
+        working-directory: src/D2L.Bmx
+        run: |
+          dotnet publish -r $env:PLATFORM-x64 -c Release -o build
+      - name: upload build
+        if: ${{ github.event_name == 'repository_dispatch' }}
+        shell: bash
         env:
+          RELEASE_TAG: ${{ github.event.client_payload.release_tag }}
           PLATFORM: ${{ matrix.platform }}
+          FILE_NAME: ${{ matrix.file_name }}
+          GH_TOKEN: ${{ github.token }}
         working-directory: src/D2L.Bmx
-        run: dotnet publish -r $env:PLATFORM-x64 -c Release -o out
+        run: |
+          cd build
+          mv "$FILE_NAME" "$PLATFORM-$FILE_NAME"
+          gh release upload "$RELEASE_TAG" "$PLATFORM-$FILE_NAME"
```

Comment:
There kind of is a worry that the create release step fails and then this tries to upload to a tag that's not created. Should just fail the workflow rather than do anything bad. Wasn't sure how to resolve this because the step has to be in this job since the file is getting built here. And we don't want to use build artifacts to upload / download at a later time

Reply:
we can make the `build` job depend on the `create_release` job.

in the `if` condition of the build job, specify `!failure()` so the build job would run even when create_release is skipped.
by default, `success()` is assumed, which would be false when a job is skipped.